data_steward/cdr_cleaner/cleaning_rules/null_invalid_foreign_keys.py

```diff
@@ -21,84 +29,215 @@
+        """
+        return len(self.get_foreign_keys(table)) > 0
+
+    def get_col_expression(self, table):
+        """
+        This method formats the column name depending of if it is a foreign key or not. If the column is a foreign key
+        it will be prefixed with the first part of the table name. If the column is not a foreign key, there will be
+        no changes to the column name.
+
+        :param table: single table in the list of affected tables
+        :return: all the formatted column names for the table
+        """
+        col_exprs = []
+        foreign_keys = self.get_foreign_keys(table)
+        for field in self.get_field_names(table):
+            if field in foreign_keys:
+                col_expr = '{x}.'.format(x=field[:3]) + field
+            else:
+                col_expr = field
+            col_exprs.append(col_expr)
+        return ', '.join(col_exprs)
+
+    def get_join_expression(self, table):
+        """
+        This method generates the LEFT_JOIN query. Only columns that are foreign keys will be
+        used in the query generation.
+
+        :param table: single table in the list of affected tables
+        :return: LEFT_JOIN query expression
+        """
+        join_expression = []
+        for key in FOREIGN_KEYS_FIELDS:
+            if key in self.get_foreign_keys(table):
+                if key == 'person_id':
+                    table_alias = cdr_consts.PERSON_TABLE_NAME
                 else:
-                    col_expr = field
-                col_exprs.append(col_expr)
-            cols = ', '.join(col_exprs)
-
-            join_expression = []
-            for key in FOREIGN_KEYS_FIELDS:
-                if key in foreign_keys_flags:
-                    if key == 'person_id':
-                        table_alias = cdr_consts.PERSON_TABLE_NAME
-                    else:
-                        table_alias = _mapping_table_for(
-                            '{x}'.format(x=key)[:-3])
-                    join_expression.append(
-                        LEFT_JOIN.format(dataset_id=dataset_id,
-                                         prefix=key[:3],
-                                         field=key,
-                                         table=table_alias))
-
-            full_join_expression = " ".join(join_expression)
-            query = dict()
-            query[cdr_consts.QUERY] = INVALID_FOREIGN_KEY_QUERY.format(
-                cols=cols,
-                table_name=table,
-                dataset_id=dataset_id,
-                project=project_id,
-                join_expr=full_join_expression)
-            query[cdr_consts.DESTINATION_TABLE] = table
-            query[cdr_consts.DISPOSITION] = bq_consts.WRITE_TRUNCATE
-            query[cdr_consts.DESTINATION_DATASET] = dataset_id
-            queries_list.append(query)
-    return queries_list
+                    table_alias = self.get_mapping_table(
+                        '{x}'.format(x=key)[:-3])
```

Comment:
I don't really know why `mapping_table` is used here for validating the foreign keys, I think there might be tacit knowledge I don't know about, @ksdkalluri perhaps you could better comment on this?

Reply:
I'm not sure why it's there either, I don't have the historical knowledge of it so I'm hoping @ksdkalluri will know why it's here.

Reply:
we use mapping tables to join on to the domain tables since we only need is the domain_ids(primary_keys) for each table which are available in _mapping_tables and its less expensive to join onto the mapping tables than to join onto the domain tables.